compiler/damlc/daml-stdlib-src/DA/BigDecimal.daml

```diff
@@ -0,0 +1,148 @@
+-- Copyright (c) 2020 The DAML Authors. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+daml 1.2
+module DA.BigDecimal (
+#ifdef DAML_UNSTABLE
+    BigDecimal
+  , RoundingMode
+  , Floor
+  , Ceiling
+  , TowardZero
+  , HalfToEven
+  , AwayFromZero
+  , HalfTowardZero
+  , HalfAwayFromZero
+  , toNumeric
+  , toBigDecimal
+  , cast
+  , compareBigDecimal
+  , add
+  , sub
+  , mul
+  , div
+  , divMod
+#endif
+) where
+
+#ifdef DAML_UNSTABLE
+import GHC.Types (BigDec, primitive)
+import GHC.Show ()
+import Prelude hiding (div)
+
+data BigDecimal r = BigDecimal BigDec deriving (Eq, Ord)
```

Comment:
Purposely including the rounding mode as a type parameter here: 
DAML writers should consciously choose a mode, casting explicitly between modes where needed, and the type checker to verify that the choices are consistent.
The DAML code can be extended to provide more convenience functions around rounding control but the mode is type-level by design.

Reply:
I understand the appeal of trying to put it at the type-level. What I don’t understand is why you then go for `add : BigDecimal a -> BigDecimal b -> BigDecimal c`. At least to me that seems to defeat the purpose of having to be explicit in the type and . I would expect to only have `(+) : BigDecimal a -> BigDecimal a - >BigDecimal a` and an operation for doing an explicit conversion.

Reply:
Agree that the `add` contradicts the intention. It should probably not be used, at the price of slightly noisier code.

Reply:
On the contrary, by decoupling the rounding mode of the result from the rounding mode of the arguments, the compiler will force anyone using the primitive `add`, `mul`, etc. functions to address rounding mode directly by explicitly choosing a rounding mode for the result. Alternatively they can use the syntactic sugar of `(+)`, `(*)`, etc. provided they ensure the arguments share a common mode.

The intermediate case which I would agree is a problem would be the defaulting case:
`prob1 : BigDecimal a -> BigDecimal b -> BigDecimal a` or
`prob2 : BigDecimal a -> BigDecimal b -> BigDecimal b`.